SOSS/dms/soss_centroids.py

```diff
@@ -0,0 +1,377 @@
+     :param yos: the oversampling factor of the image array along the y-direction.
+     :param verbose: If set True some diagnostic plots will be made.
+
+     :type image: array[float]
+     :type halfwidth: int
+     :type yos: int
+     :type verbose: bool
+
+     :returns: ytrace_max, ytrace_min, ytrace_comb - The upper edge, lower edge and center of the trace.
+     :rtype: Tuple(array[float], array[float], array[float])
+     """
+
+    dimy, dimx = image.shape
+    halfwidth = halfwidth * yos
+
+    # Create coordinate arrays.
+    xpix = np.arange(dimx)
+    ypix = np.arange(dimy)
+    _, ygrid = np.meshgrid(xpix, ypix)
+
+    # Compute windowed slopes over the columns.
+    slopevals = np.zeros_like(image)
+    for irow in range(halfwidth, dimy-halfwidth):
+
+        # Compute the window indices.
+        ymin = irow - halfwidth
+        ymax = irow + halfwidth + 1
+
+        # Get the x and y data to find the slope to.
+        datay = image[ymin:ymax, :]
+        mask = np.isfinite(datay)
+        datax = np.where(mask, ygrid[ymin:ymax, :], np.nan)  # Need to set values NaN in y to NaN in x.
+
+        # Compute the slope.
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            xmean = np.nanmean(datax, axis=0, keepdims=True)
+            ymean = np.nanmean(datay, axis=0, keepdims=True)
+
+        with np.errstate(invalid='ignore'):
+            slope = np.nansum((datax - xmean) * (datay - ymean), axis=0) / np.nansum((datax - xmean) ** 2, axis=0)
+
+        # Set slopes computed from < 3 datapoints to NaN.
+        slopevals[irow, :] = np.where(np.sum(mask, axis=0) >= 3, slope, 0.)
+
+    # Find the upper and lower bounds on the trace.
+    args = np.nanargmax(slopevals, axis=0)
+    vals = np.nanmax(slopevals, axis=0)
+    ytrace_max = np.where(vals != 0, ypix[args], np.nan)
+
+    args = np.nanargmin(slopevals, axis=0)
+    vals = np.nanmin(slopevals, axis=0)
+    ytrace_min = np.where(vals != 0, ypix[args], np.nan)
+
+    # Scan through a range of trace widths.
+    slopes_best = np.zeros_like(xpix)
+    ytrace_best = np.zeros_like(xpix)
+    widths_best = np.zeros_like(xpix)
+    for width in range(18*yos, 27*yos):
+
+        # Add the slope and its offset negative.
+        comb = slopevals - zero_roll(slopevals, -width)
+
+        # Find the maximum resulting slope.
+        args = np.nanargmax(comb, axis=0)
+        vals = np.nanmax(comb, axis=0)
+
+        # Update the best values.
+        mask = (vals > slopes_best)
+        slopes_best = np.where(mask, vals, slopes_best)
+        ytrace_best = np.where(mask, ypix[args], ytrace_best)
+        widths_best = np.where(mask, width, widths_best)
+
+    # Set the y position to NaN if the best slope was zero.
+    ytrace_best = np.where(slopes_best != 0, ytrace_best + widths_best/2., np.nan)
+    widths_best = np.where(slopes_best != 0, widths_best, np.nan)
+
+    if verbose:
+
```

Comment:
plots should always have a title (even debug ones) and/or labels

Reply:
Improved plots.